File: impls/clojure.2/src/mal/step0_repl.clj
(ns mal.step0_repl)

(defn READ [s] s)

(defn EVAL [x] x)

(defn PRINT [s] s)

(defn rep [input]
  (->
   input
   READ
   EVAL
   PRINT))

(loop []
  (print "user> ")
  (flush)
  (when-let [line (read-line)]
    (println (rep line))
    (recur)))

File: impls/js2/step0_repl.js
const READ = (str) => str;

const EVAL = (ast) => ast;

const PRINT = (str) => str;

const rep = (str) => PRINT(EVAL(READ(str)));

import * as readline from "node:readline";
import { stdin as input, nextTick, stdout as output } from "node:process";

const rl = readline.createInterface({ input, output });

const prompt = () => {
  rl.question(`user> `, (line) => {
    if (line === "") {
      console.log(`Bye!`);
      rl.close();
      return;
    }
    console.log(rep(line));
    nextTick(prompt);
  });
};

prompt();
